onecodex/stats.py

```diff
@@ -255,44 +338,188 @@ def _kruskal(
+                )
             )
 
         return AlphaDiversityStatsResults(
             test=AlphaDiversityStatsTest.Kruskal,
             statistic=result.statistic,
             pvalue=result.pvalue,
+            sample_size=len(df),
             group_by_variable=group_by_column_name,
             groups=group_names,
-            posthoc_df=posthoc_df,
+            posthoc=posthoc,
         )
 
-    def _tuplize(self, value) -> tuple:
-        if not isinstance(value, tuple):
-            if isinstance(value, list):
-                # Convert to tuple so that it's hashable in self._metadata_fetch()
-                value = tuple(value)
-            else:
-                value = (value,)
-        return value
+    def beta_diversity_stats(
+        self,
+        *,
+        group_by: str | tuple[str, ...] | list[str],
+        metric: BetaDiversityMetric = BetaDiversityMetric.BrayCurtis,
+        rank: Rank = Rank.Auto,
+        alpha: float = 0.05,
+        num_permutations: int = 999,
+    ) -> BetaDiversityStatsResults:
+        """Test for significant differences between groups of samples based on their distances.
 
-    def _assert_two_groups(self, df: pd.DataFrame, group_by_column_name: str):
-        num_groups = df[group_by_column_name].nunique()
-        if num_groups != 2:
-            raise StatsException(
-                f"`group_by` must have exactly two groups to test between after filtering (found {num_groups})."
-            )
+        Beta diversity distances between samples are computed and a PERMANOVA test is performed to
+        assess whether there are significant differences between groups of samples. Posthoc pairwise
+        PERMANOVA tests are performed if the global test is found to be statistically significant
+        and there are more than two groups.
 
-    def _get_group_names_and_alpha_values(
-        self, df: pd.DataFrame, metric: str, group_by_column_name: str
-    ) -> tuple[set[str], list[pd.Series]]:
-        group_names = set()
-        group_alpha_values = []
-        for group_name, group_df in df.groupby(group_by_column_name):
-            group_names.add(group_name)
-            group_alpha_values.append(group_df[metric])
+        Parameters
+        ----------
+        group_by : str or tuple of str or list of str
+            Metadata variable to group samples by. At least two groups are required. If `group_by`
+            is a tuple or list, field values are joined with an underscore character ("_").
+        metric : {'braycurtis', 'jaccard', 'cityblock', 'manhattan', 'aitchison', 'unweighted_unifrac', 'weighted_unifrac'}, optional
+            The beta diversity distance metric to calculate. Note that 'cityblock' and 'manhattan'
+            are equivalent metrics.
+        rank : {'auto', 'kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species'}, optional
+            Analysis will be restricted to abundances of taxa at the specified level.
```

Comment:
It looks as though, if there are any samples for which we don't have abundance estimates, we might instead be falling back to read counts. We're calculating beta diversity stats on a total sample size, rather than the sample size of just those samples with abundances.
(This is not the case for alpha diversity, where we warn that samples are excluded if they don't have abundance estimates).
I think what we should do is add a new variable to beta_diversity_stats (and also alpha_diversity_stats), similar to what we have with dataframes, with ocx_metric : {'readcount_w_children', 'readcount', 'abundance'}. (Default to abundance). Then use this for the beta diversity calculation, rather than letting beta div decide. (Open to discussion).

Reply:
I added filtering of samples with missing abundances to:

- `beta_diversity_stats()`
- `plot_pcoa()`/`plot_mds()`
- `plot_pca()`

Filtering was already being done for `plot_distance()`. If filtering happens, we show a warning to the user.

I'll create a followup issue to make this change more generally in `beta_diversity()` (it ended up being more complicated to add it there after all).